extensions/sql-database-projects/src/models/deploy/deployService.ts

```diff
@@ -0,0 +1,403 @@
+
+			this.logToOutput(constants.creatingDeploymentSettingsMessage);
+			// Create commands
+			//
+
+			await this.createCommands(mssqlFolderPath, commandsFolderPath, dockerFilePath, startFilePath, imageLabel);
+
+			this.logToOutput(constants.runningDockerMessage);
+			// Building the image and running the docker
+			//
+			const createdDockerId: string | undefined = await this.buildAndRunDockerContainer(dockerFilePath, imageName, root, profile.localDbSetting, imageLabel);
+			this.logToOutput(`Docker container created. Id: ${createdDockerId}`);
+
+
+			// Waiting a bit to make sure docker container doesn't crash
+			//
+			const runningDockerId = await this.retry('Validating the docker container', async () => {
+				return await this.executeCommand(`docker ps -q -a --filter label=${imageLabel} -q`);
+			}, (dockerId) => {
+				return Promise.resolve({ validated: dockerId !== undefined, errorMessage: constants.dockerContainerNotRunningErrorMessage });
+			}, (dockerId) => {
+				return Promise.resolve(dockerId);
+			}
+			);
+
+			if (runningDockerId) {
+				this.logToOutput(constants.dockerContainerCreatedMessage(runningDockerId));
+				return await this.getConnection(profile.localDbSetting, false, 'master');
+
+			} else {
+				this.logToOutput(constants.dockerContainerFailedToRunErrorMessage);
+				if (createdDockerId) {
+					// Get the docker logs if docker was created but crashed
+					await this.executeCommand(constants.dockerLogMessage(createdDockerId));
+				}
+			}
+
+			return undefined;
+		});
+	}
+
+	private async buildAndRunDockerContainer(dockerFilePath: string, imageName: string, root: string, profile: ILocalDbSetting, imageLabel: string): Promise<string | undefined> {
+		this.logToOutput('Building docker image ...');
+		await this.executeCommand(`docker pull ${constants.dockerBaseImage}`);
+		await this.executeCommand(`docker build -f ${dockerFilePath} -t ${imageName} ${root}`);
+		await this.executeCommand(`docker images --filter label=${imageLabel}`);
+
+		this.logToOutput('Running docker container ...');
+		await this.executeCommand(`docker run -p ${profile.port}:1433 -e "MSSQL_SA_PASSWORD=${profile.password}" -d ${imageName}`);
+		return await this.executeCommand(`docker ps -q -a --filter label=${imageLabel} -q`);
+	}
+
+	private async getConnectionString(connectionUri: string): Promise<string | undefined> {
+		const getAzdataApi = await utils.getAzdataApi();
+		//const vscodeMssqlApi = getAzdataApi ? undefined : await utils.getVscodeMssqlApi();
+		if (getAzdataApi) {
+			const connection = await getAzdataApi.connection.getConnection(connectionUri);
+			if (connection) {
+				return await getAzdataApi.connection.getConnectionString(connection.connectionId, true);
+			}
+		}
+		// TODO: vscode connections string
+
+		return undefined;
+
+	}
+
+	private async connectToDatabase(profile: ILocalDbSetting, savePassword: boolean, database: string): Promise<ConnectionResult | string | undefined> {
```

Comment:
This seems like a generic thing - not something specific to the deployService. Would suggest putting it in a common module like utils.